Add QuizQuestionsManager class

Zhi-You opened the pull request:
Fixes #58. Fixes #60, which is related to #57. Requires some feedback about the exception handling I have implemented.

Zhi-You committed:
Merge branch 'master' of https://github.com/AY2021S1-CS2113T-F12-3/tp

Zhi-You committed:
Merge branch 'master' of https://github.com/AY2021S1-CS2113T-F12-3/tp

Zhi-You committed:
Merge branch 'master' of https://github.com/AY2021S1-CS2113T-F12-3/tp

Zhi-You committed:
Merge branch 'master' of https://github.com/AY2021S1-CS2113T-F12-3/tp

Zhi-You committed:
Merge branch 'master' of https://github.com/AY2021S1-CS2113T-F12-3/tp

Zhi-You committed:
Merge branch 'master' of https://github.com/AY2021S1-CS2113T-F12-3/tp

Zhi-You committed:
Merge branch 'master' into add-QuizQuestionClass

Zhi-You committed:
Add QuizQuestionClass

Zhi-You committed:
Enhance SRP for getDescription method

Zhi-You committed:
Add exception handling for when number of quiz questions specified <= 0

kstonekuan commented on src/main/java/seedu/eduke8/exception/Eduke8Exception.java:
```diff
@@ -0,0 +1,7 @@
+package seedu.eduke8.exception;
+
+public class Eduke8Exception extends Exception {
```
Thanks for adding this but would be good to use a separate branch next time

Zhi-You replied:
Alright noted, happen to have it here because it is linked to the QuizQuestion class.

kstonekuan commented on src/main/java/seedu/eduke8/question/QuizQuestionList.java:
```diff
@@ -1,55 +1,56 @@
 import java.util.Random;
+import java.util.logging.Logger;
+
+public class QuizQuestionList {
```
Good idea to separate these responsibilities to two different classes

kstonekuan replied:
Why does this not implement DisplayableList, maybe can change the way we think of and name this class?

Zhi-You replied:
Alright I think a better fitting name would be QuizQuestionsManager as it manages the quiz questions instead of simply being a list.

kstonekuan commented on src/main/java/seedu/eduke8/question/Question.java:
```diff
@@ -44,6 +43,10 @@ public boolean wasShown() {
         return wasShown;
     }
 
+    public void markAsShown() {
```
I see what you are trying to do here but is there a reason for this class to be different from the other displayables and do we need this extra method?

Zhi-You replied:
Initially I wanted the getDescription method to only be getting the description of the question, but it is also indeed true that once getDescription method is called it is clear that the question is shown too. Will be changing this! Thanks!

kstonekuan commented on src/main/java/seedu/eduke8/question/TopicQuestionList.java:
```diff
@@ -0,0 +1,53 @@
+        this.questions = questions;
+    }
+
+    public QuizQuestionList getQuizQuestionList(int numberOfQuestionsForQuiz) throws Eduke8Exception {
```
I think you should try not to depend on this class in to create the QuizQuestionList object as this violates SRP

Zhi-You replied:
Working on it!

kstonekuan reviewed (changes requested):
Looking good, almost there

Zhi-You committed:
Revert TopicQuestionList class name to QuestionList

Zhi-You committed:
Change QuizQuestionList class to QuizQuestionsManager

Changed constructor for QuizQuestionsManager to set the quiz questions
using it.

Zhi-You committed:
Remove markAsShown method and incorporated it into getDescription()

josephhhhhhhhh commented on src/main/java/seedu/eduke8/question/QuizQuestionsManager.java:
```diff
@@ -0,0 +1,74 @@
+        if (numberOfQuestionsForQuiz > questionsInTopic.size()) {
+            throw new Eduke8Exception(INSUFFICIENT_TOPIC_QUESTIONS_FOR_QUIZ);
+        }
+
+        int numberOfQuestionsSelected = 0;
+
+        // prevent repeated questionsInTopic from being selected again
+        ArrayList<Integer> integersChosen = new ArrayList<>();
+
+        while (numberOfQuestionsSelected < numberOfQuestionsForQuiz) {
+            // get a random question that is within the bounds of the size of the available question list
+            int randomQuestionIndex = RANDOM.nextInt(questionsInTopic.size() - 1);
+
+            // if the number is already selected - the question is already selected, we re run the loop
+            // to select another random number
+            if (integersChosen.contains(randomQuestionIndex)) {
+                // logger.info("chosen a repeated question");
+                continue;
+            }
+
+            // choose a random question from the question list and add it to the questionsInTopic for quiz
+            quizQuestions.add((Question) questionsInTopic.get(randomQuestionIndex));
+            numberOfQuestionsSelected++;
+            integersChosen.add(randomQuestionIndex);
+        }
+    }
+
+    public Question getNextQuestion() {
+        // Gets current question and increment count for number of questions shown to user
+        return quizQuestions.get(currentQuestionNumber++);
+    }
+
+    public int getCurrentQuestionNumber() {
+        return currentQuestionNumber;
+    }
+
+    public boolean areAllQuestionsAnswered() {
+        return currentQuestionNumber == quizQuestions.size();
+    }
+}
```
The coding logic here looks really good! Nice work!

josephhhhhhhhh commented on src/main/java/seedu/eduke8/question/QuestionList.java:
```diff
@@ -4,88 +4,51 @@
-        ArrayList<Integer> integersChosen = new ArrayList<>();
+    /*
+    public QuizQuestionsManager getQuizQuestionList(int numberOfQuestionsForQuiz) throws Eduke8Exception {
+        QuizQuestionsManager quizQuestionsManager = new QuizQuestionsManager();
```
Since you're getting QuizQuestionList with this function, would calling the new QuizQuestionsManager object quizQuestionList instead of quizQuestionsManager?

Zhi-You replied:
Oh! Sorry! This comment is supposed to be deleted from the commit as it was from a previous version! I will change it now.

josephhhhhhhhh reviewed (commented):
I apologise if I'm nitpicking for naming convention, I'll let someone else give this a once over as well to approve the code changes. Thanks for the hard work!

Zhi-You committed:
Remove commented out code section

josephhhhhhhhh reviewed (approved):
Ok, nice the rest seems solid. I'll send this through for approval.
